### PR TITLE
Attrs class for CMAQ config

### PR DESCRIPTION
**Pull request by crdanielbusch**
- [x] Add a config file for the CMAQ set up
- [x] Move config files in `config` folder
- [x] Define a class for the CMAQ configuration object
- [x] Add validators for the CMAQ config
- [x] Run tests in checks on GitHub

### CODE REVIEW

File: src/setup_runs/cmaq/read_config_cmaq.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import datetime
+
+from attrs import define, field
+import json
+from setup_runs.config_read_functions import boolean_converter, process_date_string
+
+
+def boolean_tuple(x):
+    return (boolean_converter(x))
```
Comment (lewisjared) [inline src/setup_runs/cmaq/read_config_cmaq.py:9]: This doesn't return a tuple if that is what you intended?


The comma is important

```suggestion
    return (boolean_converter(x),)
```

File: CMAQ.config.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (lewisjared) [inline CMAQ.config.json:1]: We will need to support docker/nci configurgation files. 
Maybe this should be moved to a path like config/cmaq/config.docker.json

**Review by lewisjared — COMMENTED** [review]
Looking good. Still needs some tidying up, but its in the right direction

File: src/setup_runs/cmaq/read_config_cmaq.py (hunk 1 above)
Comment (crdanielbusch) [reply src/setup_runs/cmaq/read_config_cmaq.py:9]: I also just saw that the output is not a tuple. But apparently it works either way. I'll take a closer look at it again.

**Comment by crdanielbusch** [comment]
@lewisjared I think this is ready to be reviewed.

One thing I haven't dealt with yet is checking the formats for the configuration, so specifically: do some of the values have to be in a list or not, e.g. `["d01"]` or "d01". I can say that it works if they are all in a list and does not work if they are all not in a list. I haven't found a quick way to check the individual values as everything has to run in the Docker container and I'm not sure if it's worth the effort.

As discussed, the tests run in the ci as separate step now. I will look into running them with the integration test in parallel tomorrow.

File: src/setup_runs/config_read_functions.py [hunk 3]
```diff
@@ -207,8 +87,8 @@ def parse_config(input_str: str) -> dict[str, str | bool | int] :
         sys.exit()
 
 
-def add_environment_variables(config: dict[str, str | bool | int],
-                              environment_variables: dict[str, str]) -> dict[str, str | bool | int] :
+def add_environment_variables(config: dict[str, str | bool | int], environmental_variables: dict[str, str]) -> dict[
```
Comment (lewisjared) [inline src/setup_runs/config_read_functions.py:90]:
```suggestion
def add_environment_variables(config: dict[str, str | bool | int], environment_variables: dict[str, str]) -> dict[
```

File: src/setup_runs/config_read_functions.py [hunk 4]
```diff
@@ -227,7 +107,7 @@ def add_environment_variables(config: dict[str, str | bool | int],
     envVarsToInclude = config["environment_variables_for_substitutions"].split(',')
 
     for envVarToInclude in envVarsToInclude :
-        config[envVarToInclude] = environment_variables[envVarToInclude]
+        config[envVarToInclude] = environmental_variables[envVarToInclude]
```
Comment (lewisjared) [inline src/setup_runs/config_read_functions.py:110]: I think I fixed this in another PR. Can you find and replace environmental_variables -> environment_variables. environmental_variables isn't grammatically correct here

```suggestion
        config[envVarToInclude] = environment_variables[envVarToInclude]
```

File: src/setup_runs/wrf/read_config_wrf.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        processed configuration data.
+    """
+
+    input_str = read_config_file(filename)
+
+    config = parse_config(input_str)
+
+    # fill variables in the values with environment variables
+    # - e.g. '${HOME}' to '/Users/danielbusch'
+    config = add_environment_variables(config=config, environmental_variables=os.environ)
```
Comment (lewisjared) [inline src/setup_runs/wrf/read_config_wrf.py:152]:
```suggestion
    config = add_environment_variables(config=config, environment_variables=os.environ)
```

**Review by lewisjared — APPROVED** [review]
Nice. Can you also run ruff before merging

File: unit
[the remark this answers is not in the document]
Comment (crdanielbusch) [reply unit:1]: oops that's a `mv` command gone wrong

File: src/setup_runs/config_read_functions.py (hunk 4 above)
Comment (crdanielbusch) [reply src/setup_runs/config_read_functions.py:110]: I also thought it was fixed. I got all of them with the global find and replace now

**Comment by crdanielbusch** [comment]
> Nice. Can you also run ruff before merging

I usually go through all the files and click on Code -> Reformat Code (with the ruff plugin). Is there a better way to do this?

File: tests/unit/test_read_config.py [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+
+    assert result == expected, f"Failed to load or match JSON content for {test_input}"
+
+
+@pytest.mark.parametrize("startDate, endDate, test_id", [
+    ('2022-07-01 00:00:00 UTC', '2022-07-01 00:00:00 UTC', "test_same_day"),
+    ('2022-07-01 00:00:00 UTC', '2022-07-02 00:00:00 UTC', "test_next_day"),
+    ('2022-07-01 00:00:00 UTC', '2023-07-01 00:00:00 UTC', "test_next_year"),
+], ids=lambda test_id : test_id)
+def test_015_validator_endDate_after_startDate(startDate, endDate, test_id, cmaq_config_dict) :
```
[the remark this answers is not in the document]
Comment (crdanielbusch) [reply tests/unit/test_read_config.py:354]: I feel like I can find my failed tests faster if they have numbers. At the same time, it always takes a while to update the numbering. So I'm not sure if there is a real time benefit.

**Pull request merged**